fedimint-client/src/db.rs

```diff
@@ -262,3 +276,173 @@ impl_db_lookup!(
+        >,
+    >,
+>;
+
+/// `apply_migrations_client` iterates from the on disk database version for the
+/// client module up to `target_db_version` and executes all of the migrations
+/// that exist in the `ClientMigrationMap`, including state machine migrations.
+/// Each migration in `ClientMigrationMap` updates the database to have the
+/// correct on-disk data structures that the code is expecting. The entire
+/// process is atomic, (i.e migration from 0->1 and 1->2 happen atomically).
+/// This function is called before the module is initialized and as long as the
+/// correct migrations are supplied in `ClientMigrationMap`, the module
+/// will be able to read and write from the database successfully.
+pub async fn apply_migrations_client(
+    db: &Database,
+    kind: String,
+    target_db_version: DatabaseVersion,
+    migrations: ClientMigrationMap,
+    module_instance_id: ModuleInstanceId,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+
+    let mut global_dbtx = db.begin_transaction().await;
+    migrate_database_version(
+        &mut global_dbtx.to_ref_nc(),
+        target_db_version,
+        Some(module_instance_id),
+        kind.clone(),
+    )
+    .await?;
+
+    let disk_version = global_dbtx
+        .get_value(&DatabaseVersionKey(module_instance_id))
+        .await;
+
+    let db_version = if let Some(disk_version) = disk_version {
+        let mut current_db_version = disk_version;
+
+        if current_db_version > target_db_version {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {kind} was higher than the code database version."
+            )));
+        }
+
+        while current_db_version < target_db_version {
+            let new_states = if let Some(migration) = migrations.get(&current_db_version) {
+                info!(target: LOG_DB, "Migrating module {kind} from {current_db_version} to {target_db_version}");
+                let active_states = get_active_states(&mut global_dbtx.to_ref_nc()).await;
+                let inactive_states = get_inactive_states(&mut global_dbtx.to_ref_nc()).await;
+
+                migration(
+                    &mut global_dbtx
+                        .to_ref_with_prefix_module_id(module_instance_id)
+                        .into_nc(),
+                    module_instance_id,
+                    active_states,
+                    inactive_states,
+                )
+                .await?
+            } else {
+                warn!("Missing client db migration for version {current_db_version}");
+                None
+            };
+
+            // If the client migration returned new states, a state machine migration has
+            // occurred, and the new states need to be persisted to the database.
+            if let Some((new_active_states, new_inactive_states)) = new_states {
+                persist_active_states(&mut global_dbtx.to_ref_nc(), new_active_states).await;
```

Comment:
Where do we delete the old states?

Reply:
In this function :) Perhaps this is a bad name, will change